File: AI/ai/moves.py
##
## EPITECH PROJECT, 2025
## B-YEP-400-STG-4-1-zappy-noe.carabin
## File description:
## moves
##

def get_moves_to_tile_level1_vision(tile_index):
    """Returns a basic move list for level 1 vision (tiles 0-3)."""
    if tile_index == 0: return []
    if tile_index == 1: return ["Forward", "Left", "Forward"]
    if tile_index == 2: return ["Forward"]
    if tile_index == 3: return ["Forward", "Right", "Forward"]
    return None

def get_moves_towards_sound_direction(k_direction):
    """
    Generates a list of moves towards a sound direction K.
    K is relative to player's current facing.
    K=0: current tile, K=1: front, K=2: front-left, ..., K=8: front-right.
    """
    if k_direction == 0: return []
    if k_direction == 1: return ["Forward"]
    if k_direction == 2: return ["Forward", "Left", "Forward"]
    if k_direction == 3: return ["Left", "Forward"]
    if k_direction == 4: return ["Left", "Left", "Forward"]
    if k_direction == 5: return ["Left", "Left", "Forward"]
    if k_direction == 6: return ["Right", "Right", "Forward"]
    if k_direction == 7: return ["Right", "Forward"]
    if k_direction == 8: return ["Forward", "Right", "Forward"]
    return ["Forward"]

def execute_moves(sock, reader, moves, thread_name):
    """Executes a list of moves, checking for 'dead' after each."""
    for move in moves:
        if exit_event.is_set(): return False
        response = do_action(sock, reader, move, thread_name)
        if response == "dead": return False
        time.sleep(ACTION_DELAY)
    return True

def execute_spiral_move(sock, reader, bot_spiral_state, thread_name):
    """
    Executes the next step in a spiral pattern for food searching.
    Updates bot_spiral_state.
    Returns True if the action was successful (not "dead"), False otherwise.
    """
    action_to_take = ""

    if bot_spiral_state["leg_steps_taken"] < bot_spiral_state["leg_max_steps"]:
        action_to_take = "Forward"
        bot_spiral_state["leg_steps_taken"] += 1
    else:
        action_to_take = "Left"
        bot_spiral_state["leg_steps_taken"] = 0
        bot_spiral_state["legs_completed_at_current_length"] += 1

        if bot_spiral_state["legs_completed_at_current_length"] >= 2:
            bot_spiral_state["leg_max_steps"] += 1
            bot_spiral_state["legs_completed_at_current_length"] = 0

    safe_print(f"[{thread_name}] Food Search Spiral: Action: {action_to_take}. State: {bot_spiral_state}")
    return do_action(sock, reader, action_to_take, thread_name) != "dead"

File: ai/entity/bot/moves.cpp
/*
** EPITECH PROJECT, 2025
** B-YEP-400-STG-4-1-zappy-noe.carabin
** File description:
** moves
*/

/**
 * @file moves.cpp
 * @brief Implements movement helper functions for AI navigation.  
 * @author Jason KOENIG
 * @version 1.0
 * @date 29/06/2025
 *
 * @see AI.hpp
 */

#include "../AI.hpp"

std::string ai::entity::AI::getDirectionName(Direction dir)
{
    switch (dir) {
        case UP:
            return "Forward";
        case LEFT:
            return "Left";
        case RIGHT:
            return "Right";
        default:
            return "";
    }
}

std::vector<ai::entity::Direction> ai::entity::AI::getMovesToTileLevelVision(int idx)
{
    std::vector<Direction> moves;
    int depth = 0;
    int total = 0;

    while (true) {
        int width = 2 * depth + 1;
        if (total + width > idx)
            break;
        total += width;
        ++depth;
    }

    const int x = idx - total - depth;
    if (x < 0) {
        moves.push_back(LEFT);
        for (int i = 0; i < -x; ++i)
            moves.push_back(UP);
        moves.push_back(RIGHT);
    } else if (x > 0) {
        moves.push_back(RIGHT);
        for (int i = 0; i < x; ++i)
            moves.push_back(UP);
        moves.push_back(LEFT);
    }
    for (int i = 0; i < depth; ++i)
        moves.push_back(UP);
    return moves;
}

std::vector<ai::entity::Direction> ai::entity::AI::getMovesTowardsSoundDirection(Direction dir)
{
    switch (dir) {
        case HERE:
            return {NONE};
        case UP:
            return {UP};
        case UP_LEFT:
            return {UP, LEFT, UP};
        case UP_RIGHT:
            return {UP, RIGHT, UP};
        case LEFT:
            return {LEFT, UP};
        case RIGHT:
            return {RIGHT, UP};
        case DOWN:
            return {LEFT, LEFT, UP};
        case DOWN_LEFT:
            return {LEFT, UP, LEFT, UP};
        case DOWN_RIGHT:
            return {RIGHT, UP, RIGHT, UP};
        default:
            return {};
    }
}

bool ai::entity::AI::executeMoves(std::string &look, const std::vector<Direction> &moves)
{
    for (const auto &move : moves) {
        if (!collectAllItems(look))
            return false;
        if (move != UP && move != LEFT && move != RIGHT) {
            utils::debug::Logger &logger = utils::debug::Logger::GetInstance();
            logger.log("[Error] Unknown move " + std::to_string(move));
            return false;
        }

        const std::string move_str = getDirectionName(move);
        if (doAction(move_str) == "dead")
            return false;

        look = doAction("Look");
        if (look == "dead")
            return false;
    }
    return true;
}

bool ai::entity::AI::executeSpiralMove(SpiralState &state)
{
    std::string action_to_take;

    if (state.leg_steps_taken < state.leg_max_steps) {
        action_to_take = "Forward";
        state.leg_steps_taken++;
    } else {
        action_to_take = "Left";
        state.leg_steps_taken = 0;
        state.legs_completed_at_current_length++;

        if (state.legs_completed_at_current_length >= 2) {
            state.leg_max_steps++;
            state.legs_completed_at_current_length = 0;
        }
    }

    utils::debug::Logger &logger = utils::debug::Logger::GetInstance();
    logger.log("Search Spiral Action: " + action_to_take);
    return doAction(action_to_take) != "dead";
}
